### ISSUE TITLE
Regresión lineal para estimación de tiempo de espera en reservas

### ISSUE THREAD

**Issue by Sebastian906** [Sebastian906/Zibooka-FullStack#184]
| Campo | Valor |
|-------|-------|
| **Tipo** | `feature` |
| **Prioridad** | 🟡 High |
| **Estimación** | 1 día |
| **Depende de** | #179 |
| **Bloquea** | Ninguno |

## Descripción

Cuando un usuario reserva un libro sin stock, no tiene forma de saber cuánto tiempo deberá esperar. Implementar una regresión lineal múltiple que estime el tiempo de espera basado en la posición en la cola, categoría del libro y rotación histórica. Esto mejora significativamente la experiencia de usuario al proporcionar expectativas realistas.

## Objetivo

1. Modelo de regresión lineal que predice `wait_days` (días entre reserva y fulfillment).
2. Al crear una reserva, calcular y almacenar `estimatedWaitDays` en el documento Reservation.
3. Mostrar en el frontend "Tiempo estimado de espera: ~X días".

## Features del Modelo

| Feature | Tipo | Descripción | Fuente |
|---------|------|-------------|--------|
| `queue_position` | int | Posición en la cola (priority) | Reservation.priority |
| `category_encoded` | int | Categoría del libro (label encoded) | Product.category |
| `historical_avg_wait_days` | float | Tiempo promedio histórico de reservas de este libro | Aggregation sobre Reservation |
| `book_return_rate_30d` | float | Tasa de devolución del libro (préstamos devueltos en últimos 30d) | Aggregation sobre Loan |
| `total_active_reservations` | int | Reservas activas totales para este libro en el momento | Aggregation |
| `stock_was_zero_days` | int | Días acumulados que el libro lleva sin stock | Derivado de Product.inStock |

## Target

Variable numérica continua: `wait_days` = días entre `reservation.createdAt` y la fecha en que `status` cambió a `fulfilled`.

### 1. Clase WaitTimePredictor

**Archivo:** `zibooka-ml/app/models/wait_time_predictor.py`

```python
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import joblib
from app.models.base import BasePredictor
from app.database import mongodb

class WaitTimePredictor(BasePredictor):
    def __init__(self):
        self.model = LinearRegression()
        self.feature_columns = [
            'queue_position',
            'category_encoded',
            'historical_avg_wait_days',
            'book_return_rate_30d',
            'total_active_reservations',
            'stock_was_zero_days'
        ]
        self.label_encoder = {}
        self._metrics = {}
    
    async def _build_training_data(self) -> pd.DataFrame:
        """Construye dataset con reservas cumplidas."""
        pipeline = [
            # Solo reservas cumplidas
            { '$match': { 'status': 'fulfilled' } },
            {
                '$lookup': {
                    'from': 'products',
                    'localField': 'productId',
                    'foreignField': '_id',
                    'as': 'product'
                }
            },
            { '$unwind': '$product' },
            {
                '$project': {
                    'productId': 1,
                    'createdAt': 1,
                    'updatedAt': 1,  # Fecha de fulfillment
                    'priority': 1,
                    'category': '$product.category',
                    'inStock': '$product.inStock'
                }
            }
        ]
        
        reservations_cursor = mongodb.get_collection('reservations').aggregate(pipeline)
        reservations = await reservations_cursor.to_list(length=None)
        
        df = pd.DataFrame(reservations)
        
        # Calcular wait_days
        df['createdAt'] = pd.to_datetime(df['createdAt'])
        df['updatedAt'] = pd.to_datetime(df['updatedAt'])
        df['wait_days'] = (df['updatedAt'] - df['createdAt']).dt.days
        
        return df
    
    def _engineer_features(self, df: pd.DataFrame) -> pd.DataFrame:
        """Construye features a partir de datos crudos."""
        features = df.copy()
        
        # Label encoding para categorías
        categories = features['category'].unique()
        self.label_encoder = {cat: i for i, cat in enumerate(categories)}
        features['category_encoded'] = features['category'].map(self.label_encoder)
        
        # Calcular histórico por libro
        book_stats = features.groupby('productId').agg({
            'wait_days': 'mean',
            'priority': 'count'
        }).rename(columns={'wait_days': 'historical_avg_wait_days', 'priority': 'total_reservations'})
        
        features = features.merge(book_stats, on='productId', how='left')
        
        # stock_was_zero_days requeriría track histórico de inStock
        # Por ahora, usar 0 como default hasta implementar
        features['stock_was_zero_days'] = 0
        
        return features
    
    def train(self, training_data: pd.DataFrame) -> dict:
        features = self._engineer_features(training_data)
        
        # Filtrar outliers (wait_days > 90 son probablemente errores)
        features = features[features['wait_days'] <= 90]
        
        X = features[self.feature_columns]
        y = features['wait_days']
        
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42
        )
        
        self.model.fit(X_train, y_train)
        
        # Evaluación
        y_pred = self.model.predict(X_test)
        
        self._metrics = {
            'rmse': float(np.sqrt(mean_squared_error(y_test, y_pred))),
            'mae': float(mean_absolute_error(y_test, y_pred)),
            'r2_score': float(r2_score(y_test, y_pred)),
            'coefficients': dict(zip(self.feature_columns, self.model.coef_.tolist())),
            'intercept': float(self.model.intercept_),
            'training_samples': len(X_train),
            'test_samples': len(X_test)
        }
        
        return self._metrics
    
    def predict(self, features: np.ndarray) -> np.ndarray:
        return self.model.predict(features)
    
    def predict_single(self, queue_position: int, category: str, 
                       historical_avg: float, return_rate: float,
                       active_reservations: int, stock_zero_days: int) -> float:
        """Predice tiempo de espera para una nueva reserva."""
        cat_encoded = self.label_encoder.get(category, -1)
        
        feature_vector = np.array([[
            queue_position, cat_encoded, historical_avg, return_rate,
            active_reservations, stock_zero_days
        ]])
        
        predicted_days = self.model.predict(feature_vector)[0]
        return max(0, round(predicted_days, 1))  # No negativos
    
    def save(self, path: str) -> None:
        joblib.dump({
            'model': self.model,
            'feature_columns': self.feature_columns,
            'label_encoder': self.label_encoder,
            'metrics': self._metrics
        }, path)
    
    @classmethod
    def load(cls, path: str) -> 'WaitTimePredictor':
        data = joblib.load(path)
        predictor = cls()
        predictor.model = data['model']
        predictor.feature_columns = data['feature_columns']
        predictor.label_encoder = data['label_encoder']
        predictor._metrics = data.get('metrics', {})
        return predictor
    
    def get_feature_importance(self) -> dict:
        return dict(zip(self.feature_columns, self.model.coef_.tolist()))
```

### 2. Integración en ReservationService

**Archivo:** `server/src/reservations/reservation.service.ts`

```typescript
@Injectable()
export class ReservationService {
  constructor(
    private predictionClient: PredictionClient,
    @InjectModel(Reservation.name) private reservationModel: Model<ReservationDocument>,
    @InjectModel(Product.name) private productModel: Model<ProductDocument>,
  ) {}
  
  async addToWaitingList(userId: string, productId: string): Promise<Reservation> {
    const product = await this.productModel.findById(productId);
    if (!product) throw new NotFoundException('Producto no encontrado');
    if (product.inStock) {
      throw new BadRequestException('El libro tiene stock disponible, no necesita reserva');
    }
    
    // Calcular prioridad (siguiente en cola)
    const queueSize = await this.reservationModel.countDocuments({
      productId, status: 'pending'
    });
    const priority = queueSize + 1;
    
    // Estimar tiempo de espera
    const estimatedWait = await this.predictionClient.estimateWaitTime(
      productId, priority
    );
    
    const reservation = new this.reservationModel({
      userId,
      productId,
      priority,
      estimatedWaitDays: estimatedWait,
      status: 'pending',
      requestDate: new Date()
    });
    
    return reservation.save();
  }
}
```

### 3. Schema — Agregar estimatedWaitDays

**Archivo:** `server/src/reservations/schemas/reservation.schema.ts`

```typescript
@Schema({ timestamps: true })
export class Reservation {
  // ... campos existentes
  
  @Prop({ type: Number, default: null })
  estimatedWaitDays: number;  // Tiempo estimado en días
}
```

### 4. Frontend — Mostrar tiempo estimado

**Archivo:** `client/src/pages/ProductDetails.jsx`

```jsx
{/* En el panel de reserva */}
{reservationCreated && (
  <div className="bg-blue-50 border border-blue-200 rounded p-4 mt-4">
    <h3 className="font-semibold text-blue-800">Reserva Confirmada</h3>
    <p className="text-blue-600">
      Estás en la posición #{reservation.priority} de la cola de espera.
    </p>
    {reservation.estimatedWaitDays !== null && (
      <p className="text-blue-600 mt-2">
        ⏱️ Tiempo estimado de espera: <strong>~{reservation.estimatedWaitDays} días</strong>
      </p>
    )}
  </div>
)}
```

**Archivo:** `client/src/pages/MyReservations.jsx`

```jsx
{reservations.map(res => (
  <div key={res._id} className="border rounded p-3">
    <h3>{res.productId?.title}</h3>
    <span className="text-sm text-gray-500">Posición: #{res.priority}</span>
    {res.estimatedWaitDays && (
      <span className="text-sm text-blue-500 ml-4">
        Espera estimada: ~{res.estimatedWaitDays} días
      </span>
    )}
  </div>
))}
```

## Criterios de Aceptación

- [ ] `POST /train/wait-time` entrena y guarda el modelo de regresión lineal
- [ ] Retorna métricas: RMSE, MAE, R², coeficientes
- [ ] RMSE del modelo < 5 días en datos de prueba
- [ ] `POST /predict/wait-time` retorna `{ estimated_days, confidence_interval }`
- [ ] `ReservationService.addToWaitingList()` integra estimación
- [ ] `estimatedWaitDays` se persiste en el documento Reservation
- [ ] El frontend muestra "Tiempo estimado: ~X días" al reservar
- [ ] Si no hay modelo entrenado, mostrar mensaje genérico sin estimación
- [ ] Los valores estimados nunca son negativos (clamp a 0)

## Fórmula del Modelo

```
wait_days ≈ β₀ + β₁(queue_position) + β₂(category_encoded) 
           + β₃(historical_avg_wait_days) + β₄(book_return_rate_30d)
           + β₅(total_active_reservations) + β₆(stock_was_zero_days)
```

Donde β₁ se espera positiva (a mayor posición, mayor espera) y β₄ se espera negativa (a mayor tasa de devolución, menor espera).
